### PR TITLE
Add ListHeaderComponent and ListEmptyComponent to MetroListView

### PR DESCRIPTION
**Pull request by Minishlink**
## Motivation
The new lists have a `legacyImplementation` prop that uses `MetroListView` instead of the newer implementations (`VirtualizedList`). This is useful for certain types of lists where the new implementation does not satisfy users performance-wise. It's also useful for `react-native-web`, since it currently does not support `VirtualizedList`.
`ListHeaderComponent` and `ListEmptyComponent` props were lacking in this legacy implementation.

## Test Plan

**ListEmptyComponent**:
- [Without `legacyImplementation`](https://user-images.githubusercontent.com/82368/32144298-0675735e-bcb7-11e7-815b-971d5f4e0220.png)
- [With `legacyImplementation` - Before this commit](https://user-images.githubusercontent.com/82368/32144308-36a779b4-bcb7-11e7-8a06-3f8d54b07c99.png)
- [With `legacyImplementation` - After this commit](https://user-images.githubusercontent.com/82368/32144338-aa0db5b2-bcb7-11e7-9e4e-a74127f54ed5.png)

**ListHeaderComponent**:
- [Without `legacyImplementation`](https://user-images.githubusercontent.com/82368/32144330-83a9d676-bcb7-11e7-82cf-6f5b00c33323.png)
- [With `legacyImplementation` - Before this commit](https://user-images.githubusercontent.com/82368/32144312-44436434-bcb7-11e7-820b-d1350e8715ca.png)
- [With `legacyImplementation` - After this commit](https://user-images.githubusercontent.com/82368/32144341-ae481442-bcb7-11e7-91e3-b65ec4b81ce3.png)

## Release Notes

[GENERAL] [ENHANCEMENT] [FlatList] - Added ListHeaderComponent and ListEmptyComponent to the legacy implementation

### CODE REVIEW

File: Libraries/Lists/MetroListView.js [hunk 1]
```diff
@@ -155,6 +157,11 @@ class MetroListView extends React.Component<Props, $FlowFixMeState> {
         ref={this._captureRef}
         renderRow={this._renderRow}
         renderFooter={this.props.FooterComponent && this._renderFooter}
+        renderHeader={
+          this.state.ds.getRowCount()
```
Comment (janicduplessis) [inline Libraries/Lists/MetroListView.js]: Looking at the VirtualizedList implementation I think we still need to render the header if the list is empty. Have you tested header + empty component to make sure the behaviour is the same?

Comment (Minishlink) [reply Libraries/Lists/MetroListView.js]: Thanks for the review, and sorry for the delay, I'll check this hopefully tomorrow night.

**Comment by Minishlink** [comment]
@janicduplessis It should be better :)

File: Libraries/Lists/MetroListView.js [hunk 2]
```diff
@@ -189,10 +192,24 @@ class MetroListView extends React.Component<Props, $FlowFixMeState> {
...
+   * suppresses an error when upgrading Flow's support for React. To see the
+   * error delete this comment and run Flow. */
+  _renderEmpty = () => <this.props.ListEmptyComponent key="$empty" />;
   /* $FlowFixMe(>=0.53.0 site=react_native_fb,react_native_oss) This comment
    * suppresses an error when upgrading Flow's support for React. To see the
    * error delete this comment and run Flow. */
   _renderFooter = () => <this.props.FooterComponent key="$footer" />;
+  /* $FlowFixMe(>=0.53.0 site=react_native_fb,react_native_oss) This comment
+   * suppresses an error when upgrading Flow's support for React. To see the
+   * error delete this comment and run Flow. */
```
Comment (janicduplessis) [inline Libraries/Lists/MetroListView.js]: Do we need to add keys here to avoid the missing key warning?

Comment (Minishlink) [reply Libraries/Lists/MetroListView.js]: There are keys in the _renderHeader and _renderEmpty. Is this not sufficient?

Comment (janicduplessis) [reply Libraries/Lists/MetroListView.js]: Ohh I see, yea this is good!

File: Libraries/Lists/MetroListView.js [hunk 3]
```diff
@@ -192,7 +195,21 @@ class MetroListView extends React.Component<Props, $FlowFixMeState> {
...
   _renderFooter = () => <this.props.FooterComponent key="$footer" />;
+  /* $FlowFixMe(>=0.53.0 site=react_native_fb,react_native_oss) This comment
+   * suppresses an error when upgrading Flow's support for React. To see the
+   * error delete this comment and run Flow. */
+  _renderHeader = () => <this.props.ListHeaderComponent key="$header" />;
+  _renderHeaderAndEmpty = () => [
+    this.props.ListHeaderComponent && this._renderHeader(),
+    this.props.ListEmptyComponent &&
+      !this.state.ds.getRowCount() &&
```
[the remark this answers is not in the document]
Comment (Minishlink) [reply Libraries/Lists/MetroListView.js]: Sure, done, should I rebase ?

**Comment by janicduplessis** [comment]
Looks good! Will ship on monday.

**Comment by janicduplessis** [comment]
@facebook-github-bot shipit

**Comment by Minishlink** [comment]
@janicduplessis, the bot failed to import. I rebased. Should I do something else?

**Comment by janicduplessis** [comment]
@facebook-github-bot shipit

**Comment by janicduplessis** [comment]
Sorry I missed that :(

**Comment by 4ian** [comment]
@janicduplessis Is there any way for you to be relaunching the tests? :) The PR was about to be landed but got stuck because of the failing tests. 😢

**Comment by cpojer** [comment]
We have since removed this code altogether. I'm sorry for the long wait and for not landing your change earlier.

**Pull request closed without merging**